### ISSUE TITLE
Possible inconsistency in the calculation of evaluation scores

### ISSUE THREAD

**Issue by Davidovic** [deeplearning4j/deeplearning4j#4759]
#### Issue Description

I originally created this issue 7 days ago, but since part of the questions I then asked are now clear, I rewrote it. This should be OK I guess, as there are no comments yet.

What remains unclear to me is the calculation of the F1 score.

When the classification problem consists of 2 classes, the _eval.precision()_ and _eval.recall()_ functions return averages of the calculation for each of the classes. The _eval.f1()_ does not return an average. Instead, it returns the F1 score of one of the classes.

When the classification problem consists of 4 classes, the _eval.f1()_ function also returns an average. **Question:** Does this have a reason or explanation?

**Example with 2 classes:**
eval.precision(0)   : 0.4489795918367347
eval.precision(1)   : 0.2727272727272727
eval.precision()    : 0.3608534322820037 == average 

eval.recall(0)      : 0.7333333333333333
eval.recall(1)      : 0.1
eval.recall()       : 0.41666666666666663 == average

eval.f1(0)          : 0.5569620253164558
eval.f1(1)          : 0.14634146341463417
**eval.f1()           : 0.14634146341463417 != average**

----------------------------------------

**Example with 4 classes:**
eval.precision(0)   : 0.5
eval.precision(1)   : 0.14285714285714285
eval.precision(2)   : 0.29411764705882354
eval.precision(3)   : 0.3333333333333333
eval.precision()    : 0.3175770308123249 == average

eval.recall(0)      : 0.04
eval.recall(1)      : 0.04
eval.recall(2)      : 1.0
eval.recall(3)      : 0.08
eval.recall()       : 0.29000000000000004 == average

eval.f1(0)          : 0.07407407407407407
eval.f1(1)          : 0.0625
eval.f1(2)          : 0.45454545454545453
eval.f1(3)          : 0.12903225806451613
eval.f1()           : 0.18003794667101117 == average

**Comment by Davidovic** [comment]
In the Evaluation class I see that the fBeta() function returns an average only when there are more than 2 classes, but I don't understand why. See the relevant code in the following gist:
https://gist.github.com/Davidovic/98ad56141cb8f851c940fa609c660e48

**Comment by Davidovic** [comment]
I discussed this with @AlexDBlack on Gitter and agreed to insert that conversation here, as it explains the issue pretty well:

**Davidovic**: ...I'm still having trouble with the F1 score. Why is it that, in a two-class-classification problem, the precision and recall return an average of the calculations for each class, and the f1 doesn't?

**AlexDBlack**: it should be consistent...as for which it should be - in the binary case, it makes sense (IMO) to report the non-averaged binary for the positive (assumed class 1) class only
for non-binary case, it should report the macro-averaged metric
(note should be consistent doesn't mean it is consistent)

**Davidovic**: hmm I see what you mean, but then it is assumed that the classification problem is something like "example is X / example is NOT X". But in my case (and I suppose many cases) the classes are simply "A" and "B". So I am just as interested in the scores for A as in the scores for B. Following your reasoning, why would the precision and recall return averages and not also simply return the calculation of the "positive" (1) class?

**AlexDBlack**: (note should be consistent doesn't mean it is consistent)
it seems like a legitimate issue/concern to me, in other words

**Davidovic**: OK thanks. But for now I'd then better take the individual f1 scores for the two classes and calculate the average myself I guess

**AlexDBlack**: yeah
thanks for the issue though and following up on this
definitely something we need to get sorted, and ensure it's unambiguously documented

**Comment by Davidovic** [comment]
The more I think about it, the stranger I find the logic followed in the calculation of the f1 score in a binary classification problem.

AlexDBlack wrote: "_...it makes sense (IMO) to report the non-averaged binary for the positive (assumed class 1) class only_", but it is exactly this assumption that makes little sense to me. In fact, if I'm not mistaken, the assumption depends on the alphabetical order of the subdirectories which form my labels/classes.

When I name my subdirectories "A" and "B", the ParentPathLabelGenerator / FileSplit / ... object (not sure which part is responsible) finds one subdirectory before the other (alphabetically) and it indexes the labels as "0" and "1". This index suddenly makes DL4J become judgemental and return the F1 score for one of the two classes only.

This means that if I rename the subdirectories containing the images, changing their alphabetical order, suddenly the other class will become class "1" and will be assumed "positive" and the F1 score for that class will be returned in the evaluation.

Since the precision and recall functions do return averages in a binary classification problem, I think the f1 function should too. That would prevent such "random judgemental treatment" and be more transparent and consistent (regarding the way the various scores are calculated).

**Comment by AlexDBlack** [comment]
> but it is exactly this assumption that makes little sense to me. In fact, if I'm not mistaken, the assumption depends on the alphabetical order of the subdirectories which form my labels/classes.

I have to disagree with you, at least somewhat. I don't think it's correct or reasonable to default to reporting macro-averaged binary metrics for the binary classification case. However, you raise a legitimate point that class 0 can be the positive class, even if (by convention) class 1 more commonly is positive for binary classification.

Take a look at how scikit learn handles this for f1:
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html
1. Defaults to binary metrics - i.e., report metric for *single class only* (i.e., the positive class)
2. Defaults to class 1 as the positive
3. Allow users to override the defaults


So what I'm going to do here:
1. Add a constructor argument for the positive class (default to class 1 for positive as per scikit-learn) that only gets used for the binary case.
2. For binary case: report positive class metrics by default for f1, precision, recall, etc - *not* the macro-averaged metrics
    a. User can override by specifying null positive class (then: use averaged) *or* specify say class 0 as the positive class. User can of course use the single int arg methods to get the binary metric for any class.
3. For non-binary (multi-class) case, continue to report macro averaged metrics for the no-arg methods (like f1())
4. Update/correct javadoc to clarify exactly what is returned for the binary case under different scenarios/settings (including default)
5. Update/correct stats() method to clarify exactly what is being reported for binary vs. multi-class case

**Comment by Davidovic** [comment]
Hi @AlexDBlack ,

That sounds great, thank you so much! Also for the explanation and the comparison with scikit-learn. Even though I have no experience with scikit-learn, I guess it makes sense to apply similar logic among different libraries, as long as it's sensible of course.

I understand that returning non-averaged metrics in binary cases can be the preferred way in the majority of cases, so it's good that at least the precision, recall and f1 functions will behave in the same way and all stick to non-averaging by default. At the moment it's a mix and that got me confused and spend a lot of time on the calculator :-D

Just for my understanding: is it indeed the alphabetical order of my subdirectories that defines what becomes class 0 and class 1, as I wrote previously?

thanks again!

**Comment by AlexDBlack** [comment]
Sure, no problem. And thanks again for the issue, you've raised some quite legitimate questions here.
I'm working on a pull request to make those corrections now.

> Just for my understanding: is it indeed the alphabetical order of my subdirectories that defines what becomes class 0 and class 1, as I wrote previously?

That applies to ImageRecordReader only - which does indeed default to alphabetical for the class index assigment - but that doesn't hold in general. Other data pipelines may be different; notably, label/integer assignment can be entirely up to the user (for example when reading from a CSV).

**Comment by Davidovic** [comment]
OK thanks! Good to know.

**Comment by AlexDBlack** [comment]
Fixed here: https://github.com/deeplearning4j/deeplearning4j/pull/4802
I'll close this issue once that PR is merged.
